### PR TITLE
[FLINK-22484][table] Add built-in MAP_KEYS, MAP_VALUES, MAP_FROM_ARRA…

### PR DESCRIPTION
**Pull request by snuyanzin**
## What is the purpose of the change

The PR implements three functions from the list mentioned in FLINK-22484 : `MAP_KEYS`, `MAP_VALUES`, `MAP_FROM_ARRAYS`

## Brief change log

Implementation, tests and docs for `MAP_KEYS`, `MAP_VALUES`, `MAP_FROM_ARRAYS`

## Verifying this change

This change added tests and can be verified as follows:

* `CollectionFunctionsITCase`
## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency):  no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs / JavaDocs

### CODE REVIEW

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Expressions.java [hunk 1]
```diff
@@ -603,6 +620,15 @@ private static ApiExpression apiCallAtLeastTwoArgument(
         return new ApiExpression(unresolvedCall(functionDefinition, arguments));
     }
 
+    private static ApiExpression apiCallExactlyTwoArgument(
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Expressions.java:623]: why is this change necessary?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 2]
```diff
@@ -112,6 +114,70 @@
...
+                    .name("MAP_KEYS")
+                    .kind(SCALAR)
+                    .inputTypeStrategy(
+                            sequence(
+                                    new String[] {"input"},
+                                    new ArgumentTypeStrategy[] {logical(LogicalTypeRoot.MAP)}))
+                    .outputTypeStrategy(
+                            callContext ->
+                                    Optional.of(
+                                            DataTypes.ARRAY(
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:128]: we should try to keep `BuiltInFunctionDefinitions` class slim, therefore I would move the strategy into `TypeStrategies.SPECIFIC_FOR_MAP_KEYS`.

Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:128]: then we should also wrap the strategy into `nullable(...)` because if one of the args is null, the result should be null

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/MapFromArraysInputTypeStrategy.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * {@link InputTypeStrategy} specific for {@link BuiltInFunctionDefinitions#MAP_FROM_ARRAYS}.
+ *
+ * <p>It expects exactly two array arguments. The length of input arrays must be the same.
+ */
+@Internal
+public class MapFromArraysInputTypeStrategy implements InputTypeStrategy {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/MapFromArraysInputTypeStrategy.java:45]: shouldn't be a `sequence(logical(ARRAY), logical(ARRAY))` be enough?

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java [hunk 4]
```diff
@@ -127,7 +126,8 @@ private static void testTableApiResult(
 
     private static void testTableApiError(Table inputTable, TableApiErrorTestItem testItem) {
         try {
-            inputTable.select(testItem.expression).execute();
+            // hasNext leads to call of eval which potentially could have runtime checks
```
Comment (twalthr) [inline flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:129]: the comment is very specific for a test base of different functions, not every function might have an `eval`. how about:
`we are calling hasNext here to trigger runtime exceptions`

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+                                DataTypes.STRING())
+                        .testTableApiError(
+                                call("MAP_KEYS", $("f0"), $("f1")),
+                                "Invalid function call:\nMAP_KEYS(BOOLEAN, INT)")
+                        .testResult(
+                                map(
+                                                $("f0").cast(DataTypes.BOOLEAN()),
+                                                $("f1").cast(DataTypes.INT()))
+                                        .mapKeys(),
+                                "MAP_KEYS(map[cast(f0 as boolean), cast(f1 as int)])",
```
Comment (twalthr) [inline flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java:60]: nit: maybe all SQL keywords in capital letters?

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+                        .onFieldsWithData(null, null, "item")
+                        .andDataTypes(
+                                DataTypes.BOOLEAN().nullable(),
+                                DataTypes.INT().nullable(),
+                                DataTypes.STRING())
+                        .testTableApiError(
+                                call("MAP_KEYS", $("f0"), $("f1")),
+                                "Invalid function call:\nMAP_KEYS(BOOLEAN, INT)")
+                        .testResult(
+                                map(
```
Comment (twalthr) [inline flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java:56]: could we simplify the tests by passing the map as a data fields, then we don't need to call `map()` in every test

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+                                DataTypes.INT(),
+                                DataTypes.STRING(),
+                                DataTypes.INT(),
+                                DataTypes.STRING())
+                        .testResult(
+                                map($("f0"), $("f1"), $("f2"), $("f3")).mapKeys(),
+                                "MAP_KEYS(map[f0, f1, f2, f3])",
+                                new Integer[] {1, 2},
+                                DataTypes.ARRAY(DataTypes.INT()))
+                        .testTableApiResult(
```
Comment (twalthr) [inline flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java:83]: we are executing a lot of Flink jobs under the hood, of course testing is important but they should be well chosen, e.g. we could remove this one because it is very similar to the one above

Comment (twalthr) [reply flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java:83]: I think three to four tests for `mapKeys` should be enough no?

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+                                ImmutableMap.of(
+                                        ImmutableMap.of(1, 2), ImmutableMap.of("one", "two")),
+                                DataTypes.MAP(
+                                        DataTypes.MAP(DataTypes.INT(), DataTypes.INT()).notNull(),
+                                        DataTypes.MAP(DataTypes.STRING(), DataTypes.STRING())
+                                                .notNull()))
+                        .testResult(
+                                mapFromArrays(array($("f0")), array(map($("f1"), $("f3")))),
+                                "MAP_FROM_ARRAYS(array[f0], array[map[f1, f3]])",
+                                ImmutableMap.of(1, ImmutableMap.of("one", "two")),
```
Comment (twalthr) [inline flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java:255]: we try to avoid Guava in our code base, you can create a helper method to create the map but in this case a `Collections.singletonMap` should do the job.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+/** Implementation of {@link BuiltInFunctionDefinitions#MAP_FROM_ARRAYS}. */
+@Internal
+public class MapFromArraysFunction extends BuiltInScalarFunction {
+    public MapFromArraysFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.MAP_FROM_ARRAYS, context);
+    }
+
+    public @Nullable MapData eval(ArrayData keysArray, ArrayData valuesArray) {
+        // we rely on the casting functionality via input type strategy
```
Comment (twalthr) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java:38]: copy paste error?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        // to determine the common data type
+        if (keysArray == null || valuesArray == null) {
+            return null;
+        }
+
+        if (keysArray.size() != valuesArray.size()) {
+            throw new ValidationException(
```
Comment (twalthr) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java:45]: use `FlinkRuntimeException` as the JavaDoc of `ValidationException` says this is rather used for validation phase. not runtime.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+                    "Invalid function MAP_FROM_ARRAYS call:\n"
+                            + "The length of the keys array "
+                            + keysArray.size()
+                            + " is not equal to the length of the values array "
+                            + valuesArray.size());
+        }
+        return new MapData() {
```
Comment (twalthr) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java:52]: use a named class, because during debugging it might be useful to know what kind of MapData this is

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+import javax.annotation.Nullable;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#MAP_FROM_ARRAYS}. */
+@Internal
+public class MapFromArraysFunction extends BuiltInScalarFunction {
+    public MapFromArraysFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.MAP_FROM_ARRAYS, context);
+    }
+
+    public @Nullable MapData eval(ArrayData keysArray, ArrayData valuesArray) {
```
Comment (twalthr) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java:37]: use `@Nullable` also for the args

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the PR @snuyanzin. Let me know if you have questions.

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java (hunk 4 above)
Comment (snuyanzin) [reply flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java]: yes, agree

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java (hunk 12 above)
Comment (snuyanzin) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java]: thanks for highlighting this, 
changed for all 3 functions

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java (hunk 10 above)
Comment (snuyanzin) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java]: It seems `BuiltInFunctionTestBase` expects only `ValidationException`, in the next commit I also changed that to support `FlinkRuntimeException` in tests

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java (hunk 8 above)
Comment (snuyanzin) [reply flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java]: ok, I didn't know that it is better to avoid Guava => removed it

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java (hunk 7 above)
Comment (snuyanzin) [reply flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java]: yes, seems it is possible to decrease number of tests.
I made it in the next commit

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Expressions.java (hunk 1 above)
Comment (snuyanzin) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Expressions.java]: now I noticed that `apiCall` could be used here...
thanks

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java (hunk 6 above)
Comment (snuyanzin) [reply flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java:58]: Most of them yes, however exactly for this one no as there is a check for an invalid call e.g. wrong number of arguments

**Comment by MartijnVisser** [comment]
@snuyanzin Does it make sense to restart/rework this PR?

**Comment by snuyanzin** [comment]
several times i felt lack of that feature...
probably you're right i should give it another chance

**Comment by liuyongvs** [comment]
@snuyanzin do you rework it again. if you do not have time,I can do it

File: docs/data/sql_functions.yml [hunk 13]
```diff
@@ -617,6 +617,15 @@ collection:
   - sql: ARRAY_DISTINCT(haystack)
     table: haystack.arrayDistinct()
     description: Returns an array with unique elements. If the array itself is null, the function will return null. Keeps ordering of elements.
+  - sql: MAP_KEYS(map)
+    table: MAP.mapKeys()
+    description: Returns the keys of the map as array.
+  - sql: MAP_VALUES(map)
```
Comment (liuyongvs) [inline docs/data/sql_functions.yml:623]: Returns an unordered array containing the keys of the map like spark https://spark.apache.org/docs/latest/api/sql/index.html#map_keys
i suggest add unordered array

File: flink-python/docs/reference/pyflink.table/expressions.rst [hunk 14]
```diff
@@ -226,6 +227,8 @@ advanced type helper functions
     Expression.element
     Expression.array_contains
     Expression.array_distinct
+    Expression.map_keys
+    Expression.map_values
```
Comment (liuyongvs) [inline flink-python/docs/reference/pyflink.table/expressions.rst:232]: i suggest map_from_arrays should in same location with Expression.map_keys/Expression.map_values.
Because they belong to the same  things, they shouldn't just be in lexicographical order

File: flink-python/pyflink/table/expression.py [hunk 15]
```diff
@@ -1487,6 +1487,24 @@ def array_distinct(self) -> 'Expression':
         """
         return _binary_op("arrayDistinct")(self)
 
+    @property
+    def map_keys(self) -> 'Expression':
+        """
+        Returns the keys of the map as array.
+
```
Comment (liuyongvs) [inline flink-python/pyflink/table/expression.py:1494]: same comments Ditto add unorder

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 16]
```diff
@@ -130,6 +130,47 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
...
+                    .kind(SCALAR)
+                    .inputTypeStrategy(
+                            sequence(
+                                    new String[] {"keysArray", "valuesArray"},
+                                    new ArgumentTypeStrategy[] {
+                                        logical(LogicalTypeRoot.ARRAY),
+                                        logical(LogicalTypeRoot.ARRAY)
+                                    }))
+                    .outputTypeStrategy(nullableIfArgs(SpecificInputTypeStrategies.ARRAYS_FOR_MAP))
+                    .runtimeClass(
```
Comment (liuyongvs) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:170]: SpecificInputTypeStrategies.ARRAYS_FOR_MAP -> SpecificInputTypeStrategies. MAP_FROM_ARRAYS

File: flink-python/docs/reference/pyflink.table/expressions.rst (hunk 14 above)
Comment (snuyanzin) [reply flink-python/docs/reference/pyflink.table/expressions.rst:232]: `map_from_arrays` is not a function, is is expression defined in `pyflink/table/expressions.py` 
Expressions from there are defined in a separate section. That's the reason

File: flink-python/pyflink/table/expression.py (hunk 15 above)
Comment (snuyanzin) [reply flink-python/pyflink/table/expression.py:1494]: you never know if it is unordered or ordered.
That's why I woudn't say unordered to avoid cases that sometimes by accident it could be ordered (e.g. map of size 1) meaning violating this comment

File: docs/data/sql_functions.yml (hunk 13 above)
Comment (liuyongvs) [reply docs/data/sql_functions.yml:623]: This is just a special case. the spark logical is same. but it has unordered description

Comment (snuyanzin) [reply docs/data/sql_functions.yml:623]: From my point of Spark's description is not precise enough...
I added `No order guaranteed.`

File: flink-python/pyflink/table/expression.py (hunk 15 above)
Comment (snuyanzin) [reply flink-python/pyflink/table/expression.py:1494]: I added `No order guaranteed.`

**Comment by snuyanzin** [comment]
@MartijnVisser (I tag you since you've recently asked about this)
the PR is rebased, let's wait for the feedback

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    public @Nullable MapData eval(@Nullable ArrayData keysArray, @Nullable ArrayData valuesArray) {
+        if (keysArray == null || valuesArray == null) {
+            return null;
+        }
+
+        if (keysArray.size() != valuesArray.size()) {
+            throw new FlinkRuntimeException(
+                    "Invalid function MAP_FROM_ARRAYS call:\n"
+                            + "The length of the keys array "
+                            + keysArray.size()
```
Comment (liuyongvs) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapFromArraysFunction.java:46]: this check may not be here. the check should in compile progress instead of running process. should change input type strategy

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SpecificInputTypeStrategies.java [hunk 18]
```diff
@@ -120,6 +124,41 @@ public final class SpecificInputTypeStrategies {
...
+            callContext ->
+                    Optional.of(
+                            DataTypes.ARRAY(
+                                            callContext
+                                                    .getArgumentDataTypes()
+                                                    .get(0)
+                                                    .getChildren()
+                                                    .get(0))
+                                    .nullable());
+
```
Comment (liuyongvs) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SpecificInputTypeStrategies.java:138]: 1. should use cast  KeyValueDataType and it will throw exception eagerly
 Optional.of(
                                                            DataTypes.ARRAY(
                                                                    ((KeyValueDataType) mapType)
                                                                            .getKeyDataType()))

2. why force nullable, it may not be right

Comment (snuyanzin) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SpecificInputTypeStrategies.java:136]: Makes sense, thanks
done

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java [hunk 19]
```diff
@@ -160,6 +166,117 @@ Stream<TestSetSpec> getTestSetSpecs() {
...
+                                DataTypes.MAP(
+                                        DataTypes.STRING(), DataTypes.ARRAY(DataTypes.INT()))));
+    }
+
+    // --------------------------------------------------------------------------------------------
+
+    private static <K, V> Map<K, V> of(K k1, V v1, K k2, V v2) {
+        Map<K, V> map = new HashMap<>();
+        map.put(k1, v1);
+        map.put(k2, v2);
```
Comment (liuyongvs) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java:279]: this method is not need , you can use 
`
Collections.singletonMap(
                                                Collections.singletonMap(1, 2),
                                                Collections.singletonMap(3, 4))
`
and there are some examples in MapFunctionITCase

Comment (snuyanzin) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java:279]: This one I didn't get...
One of the usages of this method is `of(1, "one", 2, "two")` to build a map with 2 entries: `1=one` and `2=two`
If I try to replace like in your suggestion then it will be `Collections.singletonMap( Collections.singletonMap(1, "one"), Collections.singletonMap(2, "two"))` and it gives a map of 2 maps each containing one entry. And this is completely different result.

Did I miss anything?

Comment (snuyanzin) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java]: The only thing that I currently see here is usage of `org.apache.flink.util.CollectionUtil.entry` and `org.apache.flink.util.CollectionUtil.map` which I made.
However, `Collections.singletonMap` definitely does not bring benefit here since maps of size bigger than 1 are requred here

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java [hunk 20]
```diff
@@ -160,6 +167,110 @@ Stream<TestSetSpec> getTestSetSpecs() {
...
+                                "Invalid function MAP_FROM_ARRAYS call:\n"
+                                        + "The length of the keys array 3 is not equal to the length of the values array 2"),
+                TestSetSpec.forFunction(BuiltInFunctionDefinitions.MAP_FROM_ARRAYS)
+                        .onFieldsWithData(
+                                new Integer[] {1, 2},
+                                new String[] {"one", "two"},
+                                new Integer[][] {new Integer[] {1, 2}, new Integer[] {3, 4}})
+                        .andDataTypes(
+                                DataTypes.ARRAY(DataTypes.INT()),
+                                DataTypes.ARRAY(DataTypes.STRING()),
```
Comment (liuyongvs) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java:261]: mapFromArrays may add a unit test for null/null

**Comment by liuyongvs** [comment]
It looks good to me. and i left one minor comments

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java (hunk 19 above)
Comment (liuyongvs) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java]: yeap, i just comments there are examples in MapFunctionITCase.

**Review by liuyongvs — APPROVED** [review]

**Comment by snuyanzin** [comment]
rebased to resolve conflicts

**Comment by chenshaojin** [comment]
> several times i felt lack of that feature... probably you're right i should give it another chance

There is a bug when using MAP_FROM_ARRAYS...

**Comment by twalthr** [comment]
@chenshaojin please open a new issue in this case.
